Add support for DirectResponse in agw

corsairier opened the pull request:
# Description
Closes: #11843 

# Change Type

Select one or more of the following by including the corresponding slash-command:
```
/kind new_feature
```

# Changelog

```release-note
Added DirectResponse Support in agentgateway
```

# Additional Notes

Tested with AgentGateway
1. Setup the kind cluster and build images using this command `./hack/kind/setup-kind.sh`
2. Install CRDs and kgateway control plane using 

```
helm upgrade -i -n kgateway-system kgateway-crds _test/kgateway-crds-1.0.0-ci1.tgz --version 1.0.0-ci1 \
          --create-namespace

helm upgrade -i -n kgateway-system kgateway _test/kgateway-1.0.0-ci1.tgz --version 1.0.0-ci1 \
          --set image.registry=ghcr.io/kgateway-dev \
          --set agentGateway.enabled=true \
          --create-namespace
```
4. Applying DirectResponse config

```
kubectl apply -f- <<EOF
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: agent-gateway
  namespace: kgateway-system
spec:
  gatewayClassName: agentgateway
  listeners:
  - name: http
    protocol: HTTP
    port: 80
    allowedRoutes:
      namespaces:
        from: All
---
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: direct-response-route
  namespace: kgateway-system
spec:
  parentRefs:
  - name: agent-gateway
    namespace: kgateway-system
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /direct
    filters:
    - type: ExtensionRef
      extensionRef:
        group: gateway.kgateway.dev
        kind: DirectResponse
        name: test-direct-response
---
apiVersion: gateway.kgateway.dev/v1alpha1
kind: DirectResponse
metadata:
  name: test-direct-response
  namespace: kgateway-system
spec:
  status: 200
  body: "Direct response from agent gateway"
EOF
```

4. Port Forwarding: 
```
kubectl port-forward -n kgateway-system deployment/agent-gateway 8080:80
```

5. Send request:

```
> curl -v localhost:8080/direct
* Host localhost:8080 was resolved.
* IPv6: ::1
* IPv4: 127.0.0.1
*   Trying [::1]:8080...
* Connected to localhost (::1) port 8080
> GET /direct HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/8.5.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< content-length: 34
< date: Thu, 07 Aug 2025 09:53:23 GMT
< 
* Connection #0 to host localhost left intact
Direct response from agent gateway
```

corsairier commented:
Hi @npolshakova,  
I’ve opened this draft PR and wanted to get some early feedback. I haven’t written E2E tests or other related tests yet, and I’m not sure if there are additional steps or requirements I might be missing. Could you please review my changes and let me know:

- If the implementation and approach are correct
- What tests or documentation should be added
- Any other improvements or requirements before this can move out of draft

Thank you for your guidance!

npolshakova commented:
Hi @corsairier! 👋  Thanks for the contribution! 

I would add the following tests:
- unit test for the filter logic: https://github.com/kgateway-dev/kgateway/blob/main/internal/kgateway/agentgatewaysyncer/route_collections_test.go#L1552
- translator-level test for the syncer: https://github.com/kgateway-dev/kgateway/blob/main/internal/kgateway/agentgatewaysyncer/agentgateway_translator_test.go 
- env test: https://github.com/kgateway-dev/kgateway/blob/main/internal/kgateway/setup/agentgateway_test.go 

With that coverage we probably don't need a dedicated e2e test, but I would write the tests you manually took to validate in the PR description so we can have an example for future docs! 

We'll also want to consider the case when there's already a route action set, similar how the envoy plugin does it: https://github.com/kgateway-dev/kgateway/blob/911f7a02c4fee803b535bd53152de851e00b00e2/internal/kgateway/extensions2/plugins/directresponse/direct_response_plugin.go#L121

corsairier commented:
Thanks for your feedback, @npolshakova! I'll make those changes!

corsairier commented:
@npolshakova, I have added the following tests,


> * unit test for the filter logic: https://github.com/kgateway-dev/kgateway/blob/main/internal/kgateway/agentgatewaysyncer/route_collections_test.go#L1552
> * translator-level test for the syncer: https://github.com/kgateway-dev/kgateway/blob/main/internal/kgateway/agentgatewaysyncer/agentgateway_translator_test.go

Also I have added the logic for this edge case,
> We'll also want to consider the case when there's already a route action set, similar how the envoy plugin does it:
> 
> https://github.com/kgateway-dev/kgateway/blob/911f7a02c4fee803b535bd53152de851e00b00e2/internal/kgateway/extensions2/plugins/directresponse/direct_response_plugin.go#L121

However, I am not sure about how to implement the env test. Could you provide some guidance or examples on how to approach writing env test here?

npolshakova commented:
Yep, the agentgateway env tests are run here: https://github.com/kgateway-dev/kgateway/blob/main/internal/kgateway/setup/agentgateway_test.go 

The test files live here where you can find some samples: https://github.com/kgateway-dev/kgateway/tree/main/internal/kgateway/setup/testdata/agentgateway. These inputs are the same as the translator-level testdata, but the env tests check that the xds is being correctly served.

Once you create an input file, if you run the test it should output a file with the config dump that you can check looks as expected with the direct response route.

corsairier commented:
Hi @npolshakova,
Thanks for your guidance. This PR is now ready for review. I have added the env tests along with the tests I manually took in the PR description.

npolshakova commented on internal/kgateway/agentgatewaysyncer/conversion.go:
```diff
@@ -110,6 +108,10 @@ func convertHTTPRouteToADP(ctx RouteContext, r gwv1.HTTPRouteRule,
+	if filterError != nil {
+		return res, filterError
+	}
 	return res, backendErr
```
Let's do a multi error here (create a list of errs, append and then errors.Join(errs...)

npolshakova commented:
@corsairier looks like there are some merge conflicts. Let me know when they're resolved and I can re-approve!

corsairier commented on internal/kgateway/setup/testdata/agentgateway/grpcroute-out.json:
```diff
@@ -22,6 +22,27 @@
         }
       }
     },
+    {
```
Would you recommend running the tests individually using TEST_PREFIX? This removed the direct response service for me

npolshakova replied:
https://github.com/kgateway-dev/kgateway/issues/11968

npolshakova commented:
@corsairier looks like there's a merge conflict. Let me know when it's resolved and I can re-approve.

corsairier commented:
@npolshakova, Now its resolved
